### PR TITLE
Start of a Chunk Manager

### PR DESCRIPTION
**Pull request by dualspiral**
**SpongeAPI** | [Sponge](https://github.com/SpongePowered/Sponge/pull/3228)

Creates a `ChunkManager`, currently just fronts chunk loading tickets.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/world/server/TicketType.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+@CatalogedBy(TicketTypes.class)
+public interface TicketType<T> extends CatalogType {
+
+    /**
+     * Gets the lifetime of any {@link Ticket tickets} of this type.
+     *
+     * @return The number of {@link Ticks} any {@link Ticket tickets} of this
+     *         type will be valid for.
+     */
+    Ticks getTicketLifetime();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/TicketType.java]: `getLifetime()`?

File: src/main/java/org/spongepowered/api/world/server/ChunkManager.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import org.spongepowered.api.world.chunk.Chunk;
+import org.spongepowered.math.vector.Vector3i;
+
+import java.util.Optional;
+
+/**
+ * Manages {@link Chunk chunks} for a {@link ServerWorld}.
+ */
+public interface ChunkManager {
+
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/ChunkManager.java:41]: `ServerWorld getWorld()`?

File: src/main/java/org/spongepowered/api/world/server/ChunkManager.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+     * @param ticket The ticket to check.
+     * @return true if so
+     */
+    boolean isValid(Ticket<?> ticket);
+
+    /**
+     * Gets the default lifetime, in {@link Ticks}, of any ticket requested via
+     * this ticket manager.
+     */
+    Ticks getDefaultTicketLifetime(TicketType<?> ticketType);
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/ChunkManager.java]: `getDefaultLifetime()`

File: src/main/java/org/spongepowered/api/world/server/TicketTypes.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.function.Supplier;
+
+/**
+ * Types of {@link Ticket tickets} that can be requested via the
+ * {@link ChunkManager}.
+ */
+public class TicketTypes {
+
+    /**
+     * Represents {@link Ticket tickets} that forces a chunk to stay loaded.
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/TicketTypes.java]: that force chunk(s) ...

File: src/main/java/org/spongepowered/api/world/server/ChunkManager.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Request a {@link Ticket} for the given {@link TicketType}.
+     *
+     * @param type The type of ticket to request.
+     * @param value The value to register the ticket with.
+     * @param <T> The type of the supplied {@code value}.
+     * @return The ticket, if granted.
+     */
+    <T> Optional<Ticket<T>> requestTicket(TicketType<T> type, Vector3i chunkPosition, T value);
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/ChunkManager.java]: Radius?

File: src/main/java/org/spongepowered/api/world/server/ChunkManager.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+     * Releases the provided {@link Ticket}, allowing the chunk position
+     * represented by the given ticket to be unloaded (if it is not being kept
+     * loaded by other means).
+     *
+     * @param ticket The ticket to release.
+     */
+    boolean releaseTicket(Ticket<?> ticket);
+
+}
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/ChunkManager.java:171]: `<T> Collection<Ticket<T>> getTickets(TicketType<T> type)` ?

File: src/main/java/org/spongepowered/api/world/server/ServerWorld.java [hunk 7]
```diff
@@ -203,4 +203,12 @@ default Chunk getChunkAtBlock(int bx, int by, int bz) {
      * @return The raid at that location, if present
      */
     Optional<Raid> getRaidAt(Vector3i blockPosition);
+
+    /**
+     * Gets the {@link ChunkManager} for this world.
+     *
+     * @return The ticket manager.
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/ServerWorld.java]: the chunk manager

File: src/main/java/org/spongepowered/api/world/server/Ticket.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.world.server;
+
+/**
+ * Represents a chunk loading ticket.
+ *
+ * @param <T> The type of value this ticket holds.
+ */
+public interface Ticket<T> {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/Ticket.java:32]: `TicketType<T> getType()`

File: src/main/java/org/spongepowered/api/world/server/TicketTypes.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import org.spongepowered.api.Sponge;
+import org.spongepowered.math.vector.Vector3i;
+
+import java.util.function.Supplier;
+
+/**
+ * Types of {@link Ticket tickets} that can be requested via the
+ * {@link ChunkManager}.
+ */
+public class TicketTypes {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/world/server/TicketTypes.java]: `final`

File: src/main/java/org/spongepowered/api/world/server/ChunkManager.java (hunk 5 above)
Comment (Zidane) [reply src/main/java/org/spongepowered/api/world/server/ChunkManager.java]: Additional, the chunkPosition really is the chunkOrigin that the radius operates off of.

File: src/main/java/org/spongepowered/api/world/server/ChunkManager.java (hunk 3 above)
Comment (dualspiral) [reply src/main/java/org/spongepowered/api/world/server/ChunkManager.java]: Actually just removed this, it's on the `TicketType` instead.

**Comment by pop4959** [comment]
Zidane asked me to look over this, and I'll say that I think it looks pretty nice so far.

However, there doesn't seem to be any way for a plugin to be notified as to when the chunk(s) for a particular ticket have been loaded, at a given chunk state (if not just loaded vs generated). I think this would be a worthwhile addition as many plugins will not care about the tickets themselves, but really just want access to chunks. In API 7 this would have been possible using something such as World#loadChunkAsync since it returns a future (and yes, I'm aware that API was never really functional, but it gets the point I'm trying to make across).

File: src/main/java/org/spongepowered/api/registry/RegistryTypes.java [hunk 10]
```diff
@@ -367,6 +368,8 @@
 
     public static final RegistryType<TeleportHelperFilter> TELEPORT_HELPER_FILTER = RegistryTypes.spongeKey("teleport_helper_filter");
 
+    public static final RegistryType<TicketType<?>> TICKET_TYPE = RegistryTypes.spongeKey("ticket_types");
```
Comment (Lignium) [inline src/main/java/org/spongepowered/api/registry/RegistryTypes.java]: ticket_types -> ticket_type

File: src/main/java/org/spongepowered/api/world/server/ChunkManager.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+     * Checks if the provided {@link Ticket} is valid for the world this manager
+     * represents.
+     *
+     * @param ticket The ticket to check.
+     * @return true if so
+     */
+    boolean isValid(Ticket<?> ticket);
+
+    /**
+     * Gets the {@link Ticks} remaining on the supplied ticket.
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/world/server/ChunkManager.java:59]: What's the behavior of a ticket that isn't valid?

Comment (dualspiral) [reply src/main/java/org/spongepowered/api/world/server/ChunkManager.java:59]: Doesn't do anything. The distance manager has exponged all knowledge of it so the chunks it refers to may unload.

**Comment by dualspiral** [comment]
For anyone who is even _thinking_ about merging, please don't. While I think I've gotten most of the issues I'm still in the middle of testing, chunk tickets are a little weird.

> However, there doesn't seem to be any way for a plugin to be notified as to when the chunk(s) for a particular ticket have been loaded, at a given chunk state (if not just loaded vs generated). I think this would be a worthwhile addition as many plugins will not care about the tickets themselves, but really just want access to chunks. In API 7 this would have been possible using something such as World#loadChunkAsync since it returns a future (and yes, I'm aware that API was never really functional, but it gets the point I'm trying to make across).

I've actually spent a lot of time looking at this and it's not going to happen, at least short term. The problem is that the implementation doesn't really hold a link between a ticket and the chunk loading lifecycle so much - the ticket and actual action are so far removed from each other that a ticket is nothing more than a level to Minecraft - there is no "completion" notification.

The best I'll be able to do (and I'm not doing it as part of this) will be to just have our tickets listen for a chunk load and notify based on that - but there is no guarantee that a chunk load would happen as part of a ticket.

Getting the tickets working with the manager as we expect is the goal for this PR. There is no reason why such a mechanism, should one be found, couldn't be added in a non-breaking change manner to a future revision of API 8 if it misses the original cut.

**Comment by pop4959** [comment]
Thanks for taking a look. I agree it probably doesn't make too much sense as part of the ticket system for the reasons you've just described. The original idea I had was that a loadChunk method returning a future could be re-added in API 8 (as mentioned in #2287). Implementations of this already exist on other platforms, though another mechanism could also be used. Perhaps after some further discussion this can be considered for a revision of API 8 (since the freeze is approaching).

**Pull request merged**